File: python/compresr/agents/facades/anthropic.py
"""Anthropic-shaped facade — ``client.messages.create(...)``.

Delegates the actual run to a private :class:`_Engine` and remaps the
returned :class:`NormalizedResult` to an Anthropic ``Message`` shape.
"""

from __future__ import annotations

from typing import Any, Optional

from ..schemas.anthropic import AnthropicMessage, to_anthropic_message


def _forward(kw: dict, **maybe: Any) -> dict:
    """Return ``kw`` with each non-``None`` ``maybe`` value added.

    Used to keep explicit forwarding kwargs out of the engine call when
    the caller didn't set them, so engine-side defaults still apply.
    """
    for key, value in maybe.items():
        if value is not None:
            kw[key] = value
    return kw


class _AnthropicMessages:
    """``client.messages`` surface — exposes ``.create`` / ``.acreate``."""

    def __init__(self, engine: Any) -> None:
        self._engine = engine

    def create(
        self,
        *,
        model: Optional[str] = None,
        messages: list,
        max_tokens: int = 4096,
        tools: Optional[list] = None,
        system: Optional[Any] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None,
        top_k: Optional[int] = None,
        stop_sequences: Optional[list] = None,
        **kw: Any,
    ) -> AnthropicMessage:
        forwarded = _forward(
            dict(kw),
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop_sequences=stop_sequences,
        )
        result = self._engine.run(
            messages=messages,
            tools=tuple(tools or ()),
            system=system,
            max_tokens=max_tokens,
            model=model,
            **forwarded,
        )
        # The engine raised if no effective model could be resolved, so
        # ``model or default`` is safe here for the response envelope.
        effective_model = model or self._engine.default_model_name
        return to_anthropic_message(result, model=effective_model)

    async def acreate(
        self,
        *,
        model: Optional[str] = None,
        messages: list,
        max_tokens: int = 4096,
        tools: Optional[list] = None,
        system: Optional[Any] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None,
        top_k: Optional[int] = None,
        stop_sequences: Optional[list] = None,
        **kw: Any,
    ) -> AnthropicMessage:
        forwarded = _forward(
            dict(kw),
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop_sequences=stop_sequences,
        )
        result = await self._engine.arun(
            messages=messages,
            tools=tuple(tools or ()),
            system=system,
            max_tokens=max_tokens,
            model=model,
            **forwarded,
        )
        effective_model = model or self._engine.default_model_name
        return to_anthropic_message(result, model=effective_model)


class _Anthropic:
    """``client`` surface — exposes ``.messages``."""

    def __init__(self, engine: Any) -> None:
        self.messages = _AnthropicMessages(engine)


__all__ = ["_Anthropic", "_AnthropicMessages"]

File: typescript/src/agents/facades/anthropic.ts
/**
 * Anthropic-shaped facade — ``client.messages.create(...)``.
 *
 * Delegates the actual run to a private :class:`CompresrEngine` and remaps the
 * returned :class:`NormalizedResult` into an Anthropic ``Message`` shape.
 *
 * Mirrors Python ``compresr/agents/facades/anthropic.py``.
 */
import { CompresrError } from '../../errors/index.js';
import { CompresrEngine } from '../engine.js';
import {
  toAnthropicMessage,
  type AnthropicMessage,
} from '../schemas/anthropic.js';

export interface AnthropicMessagesCreateOptions {
  /**
   * Optional per-call model. Defaults to the model encoded on the engine's
   * ``llm`` spec (``'anthropic:claude-...'``). If neither is set, the engine
   * throws ``CompresrError("missing_model")``.
   */
  model?: string;
  messages: ReadonlyArray<unknown>;
  maxTokens?: number;
  tools?: ReadonlyArray<unknown>;
  system?: string;
  config?: Record<string, unknown>;
  temperature?: number;
  topP?: number;
  topK?: number;
  stopSequences?: string[];
}

export interface AnthropicMessagesFacade {
  create(options: AnthropicMessagesCreateOptions): Promise<AnthropicMessage>;
  /** Streaming not yet implemented — throws. */
  stream(options: AnthropicMessagesCreateOptions): Promise<never>;
}

export function anthropicMessages(engine: CompresrEngine): AnthropicMessagesFacade {
  return {
    async create(options: AnthropicMessagesCreateOptions): Promise<AnthropicMessage> {
      const runOpts: Parameters<typeof engine.run>[0] = {
        messages: options.messages,
        tools: options.tools ?? [],
        maxTokens: options.maxTokens ?? 4096,
      };
      if (options.model !== undefined) runOpts.model = options.model;
      if (options.system !== undefined) runOpts.system = options.system;
      if (options.config !== undefined) runOpts.config = options.config;
      if (options.temperature !== undefined) runOpts.temperature = options.temperature;
      if (options.topP !== undefined) runOpts.topP = options.topP;
      if (options.topK !== undefined) runOpts.topK = options.topK;
      if (options.stopSequences !== undefined) {
        runOpts.stopSequences = options.stopSequences;
      }
      const result = await engine.run(runOpts);
      const responseModel = options.model ?? engine.defaultModelName ?? '';
      return toAnthropicMessage(result, { model: responseModel });
    },
    stream(_options: AnthropicMessagesCreateOptions): Promise<never> {
      return Promise.reject(
        new CompresrError(
          'streaming not yet implemented — Phase 2 work item',
          'not_implemented'
        )
      );
    },
  };
}
